container/lru/map.go

```diff
@@ -0,0 +1,524 @@
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	// Nothing can be added when the limit is zero.
+	if m.limit == 0 {
+		return 0
+	}
+
+	// Enable item expiration when not already done and the passed TTL is
+	// non-zero.  Note that this being true also implies there is no default TTL
+	// configured since item expiration would already be marked active in that
+	// case.
+	//
+	// Otherwise, scan through the items and remove any that are expired when
+	// item expiration was already enabled and the scan interval has elapsed.
+	// This is done for efficiency so the scan only happens periodically instead
+	// of on every put.
+	now := m.nowFn()
+	if !m.hasTimeouts && ttl > 0 {
+		m.hasTimeouts = true
+		m.nextExpireScan = now.Add(expireScanInterval)
+	} else if m.hasTimeouts && now.After(m.nextExpireScan) {
+		numEvicted = m.removeExpired(now)
+	}
+
+	numEvicted += m.put(key, value, now, ttl)
+	return numEvicted
+}
+
+// Get returns the value associated with the passed key if it is a member and
+// modifies its priority to be the most recently used item.  The expiration time
+// for the item is not modified.
+//
+// The boolean return value indicates whether or not the item lookup was
+// successful.
+//
+// See [Peek] for a variant that does not modify the priority or update the hit
+// ratio.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Get(key K) (V, bool) {
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	elem, exists := m.items[key]
+	if !exists || m.isElemExpired(elem, m.nowFn()) {
+		m.misses++
+		return *new(V), false
+	}
+
+	m.hits++
+	m.moveElemToFront(elem)
+	return elem.value, true
+}
+
+// Exists returns whether or not the passed item is a member.  The priority and
+// expiration time for the item are not modified.  It does not affect the hit
+// ratio.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Exists(key K) bool {
+	m.mtx.Lock()
+	elem, exists := m.items[key]
+	exists = exists && !m.isElemExpired(elem, m.nowFn())
+	m.mtx.Unlock()
+	return exists
+}
+
+// Peek returns the associated value of the passed key if it is a member without
+// modifying any priority or the hit ratio.  The boolean return value indicates
+// whether or not the item lookup is successful.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Peek(key K) (V, bool) {
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	elem, exists := m.items[key]
+	if !exists || m.isElemExpired(elem, m.nowFn()) {
+		return *new(V), false
+	}
+
+	return elem.value, true
+}
+
+// Len returns the number of items in the map.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Len() uint32 {
+	m.mtx.Lock()
+	numItems := uint32(len(m.items))
+	m.mtx.Unlock()
+	return numItems
+}
+
+// Delete deletes the item associated with the passed key if it exists.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Delete(key K) {
```

Comment:
Given a Map configured with no default ttl, when `PutWithTTL` is called it will set `m.hasTimeouts` to true.

If that element is removed via `Delete`, nothing remaining in the Map will have a ttl but `m.hasTimeouts` will still be set. 

afaict this applies to every place where an element is removed from the Map.

Reply:
Yes, you are correct and it is intentional.  The primary purpose of the flag is to efficiently optimize the common usage pattern of creating a new map without any expiration while never adding any items with expiration.  There is no efficient way to update the flag when an item is removed because it would require scanning _all_ items on every single item eviction/removal to determine if anything still has a timeout in order to toggle it and that would defeat the purpose of the optimization to begin with.

The rationale is that, in practice, if a timeout is ever used, it is exceedingly likely more things will use a timeout in the future too.  Importantly, the flag being set does not break behavior at all, it just means the periodic expire scans won't be avoided.

I probably should rename the flag to `probablyHasTimeouts` to make it clear that it is intentional.

Reply:
Renamed and added a comment:

> // probablyHasTimeouts is used to optimize out expiration scans when there
> // is no default TTL and there have never been any items with a TTL.